doc/modified.manpages/boot.man
.TH BOOT 8
.SH NAME
boot \- transfer control from the kernel to rc
.SH SYNOPSIS
.B /boot/boot
[
.I args
]
.SH DESCRIPTION
.I Boot
is the first program run after a kernel has been loaded.
It sets up usb input devices if needed and creates a minimal namespace.
It then
.IR exec (2)'s
the
.IR plan9rc (8)
script.
It is started by the kernel, never run directly by the user.  See
.IR booting (8)
for information about the process of loading the kernel (and
.IR boot )
into memory.
.PP
Once loaded, the kernel initializes its data structures and devices.
It sets the two environment variables
.B /env/cputype
and
.B /env/terminal
to describe the processor.
It then binds a place-holder file server,
.IR root (3),
onto
.B /
and crafts an initial process whose sole function is to
.IR exec (2)
.BR /boot/boot ,
a binary which is compiled into
.IR root (3).
.I Boot
completes by
.IR exec (2)'ing
.B /boot/plan9rc.
.PP
The following information applies mostly to 
.IR plan9rc (8)
but has been left in this section
of the manual for convenience to existing users. In general read "plan9rc" for "boot"
in the section that follows and consult the plan9rc documentation for definitive reference.
.SS CPU Servers
The user owning devices and console processes on CPU servers
and that user's domain and encryption key are
read from NVRAM on all machines except PC's.
PC's keep the information in the disk partition
.BI /dev/sd XX /nvram\f1.
If a
.B -k
option is given or if no stored information is found
.I boot
will prompt for all three items and store them.
.IP
.EX
password:
authid: bootes
authdom: research.bell-labs.com
.EE
.PP
The key is used for mutual authentication of the server and its clients.
The domain and id identify the owner of the key.
.PP
.SS Booting Methods
The methods available to any system depend on what was
compiled into the kernel.
The complete list of booting methods are listed below.
.TP 8
.BR tcp
connect via Ethernet using the TCP protocol.
The
.I args
are passed to
.IR ipconfig (8)
when configuring the IP stack.
The
.IR plan9.ini (8)
variables
.B fs
and
.B auth
override the file server and authentication server IP addresses
obtained (if any) from DHCP during
.IR ipconfig (8).
.TP 8
.B local
connect to the local file system.
The first argument is a disk holding a file system.
.I Boot
inspects the disk.
If the disk is a
.IR fossil (4)
file system, it invokes
.B /boot/fossil
to serve it.
If the
.B venti
environment variable (really,
.IR plan9.ini (8)
variable) is set,
.I boot
first arranges for fossil to be able to
contact the named
.IR venti (8)
server.
The variable's value can take the following forms:
.RS
.TP
.B /dev/sdC0/arenas
the file should be a venti partition with a configuration
stored on it using
.I venti/conf
(see
.IR venti-fmt (8)).
.I Boot
will start a loopback IP interface on 127.0.0.1
and start
.I venti
announcing on
.B tcp!127.1!17034
for venti service
and
.B tcp!127.1!8000
for web service,
using the configuration stored in that partition.
.TP
.B /dev/sdC0/arenas tcp!*!17034
same as the last but specify an alternate venti service address.
In this example, using
.B *
will announce on all available IP interfaces (even ones configured later)
rather than just the loopback device.
The loopback interface is still configured.
.TP
.B /dev/sdC0/arenas tcp!*!17034 tcp!*!80
same as the last but specify alternate venti service and web addresses.
The loopback interface is still configured.
.TP
.B tcp!135.104.9.2!17034 \fR[ \fIargs\fR ]
the network address of a venti server running on a separate machine.
.I Boot
will configure the IP stack by passing
.IR args ,
if any, to
.IR ipconfig (8).
.RE
.IP
If the disk is not a
.IR fossil (4)
partition,
.I boot
invokes
.BR /boot/kfs .
A variety of programs, like
.I 9660srv
and
.IR dossrv (4)
masquerade as
.I kfs
to allow booting from alternate media formats,
so as long as the disk is not a
.I fossil
disk, no check is made that the disk is in fact
a
.I kfs
disk.
The args are passed to
.IR kfs (4).
.IP
For the
.B tcp
method,
the address must be a numeric IP address.
If no address is specified,
a file server address will be found from another
system on the network using the BOOTP protocol and
the Plan 9 vendor-specific fields.
.SH EXAMPLES
On PCs, the default arguments to boot are constructed using
the
.B bootargs
variable in
.IR plan9.ini (8).
.PP
Start
.IR kfs (4)
with extra disk buffers:
.IP
.EX
bootargs=local!#S/sdC0/fs -B 4096
.EE
.LP
Use an IP stack on an alternate ethernet interface
with a static address and fixed file server and authentication
server addresses.
.IP
.EX
fs=192.168.0.2
auth=192.168.0.3
bootargs=tcp -g 192.168.0.1 ether /net/ether1 \e
    192.168.0.50 255.255.255.0
.EE
.LP
(The
.B bootargs
line is split only for presentation; it is one line in the file.)
.SH FILES
.B #s/boot
.br
.B #//boot/boot
.SH SOURCE
.B /sys/src/9/boot
.SH "SEE ALSO"
.IR root (3),
.IR factotum (4),
.IR dhcpd (8),
.IR plan9rc (8)
.SH BUGS
The use of
.B bootargs
in general is odd.
The configuration specification
for fossil and venti servers
is particularly odd, but it does
cover the common cases well.


src.9.boot/boot.c
#include <u.h>
#include <libc.h>
#include <fcall.h>
#include "../boot/boot.h"

/* This is a simplified and conjoined version of boot and init designed for 'rootless' bootup*/


#define PARTSRV "partfs.sdXX"

enum {
	Dontpost,
	Post,
};

/* vars and functions from standard boot.c */
char	cputype[64];
char	sys[2*64];
char 	reply[256];
int	printcol;
int	mflag;
int	fflag;
int	kflag;

char *rdparts;
char	*bargv[Nbarg];
int	bargc;

void boot(int argc, char *argv[]);
static void	swapproc(void);
static void	kbmap(void);

/* vars and functions from standard init.c */
char*	readenv(char*);
void	settheenv(char*, char*);
void	cpenv(char*, char*);
void	closefds(void);
void	fexec(void(*)(void));
void	rcexec(void);
void	plan9start(void);
void init(void);
void pinhead(void*, char *msg);

char	*service;
char	*cmd;
char	*cpu;
char *checkmanual;
int	manual;

void
boot(int argc, char *argv[])
{
	int fd;

	fmtinstall('r', errfmt);

	/*
	 * we should inherit the standard fds all referring to /dev/cons,
	 * but we're being paranoid.
	 */
	close(0);
	close(1);
	close(2);
	bind("#c", "/dev", MREPL);
	open("/dev/cons", OREAD);
	open("/dev/cons", OWRITE);
	open("/dev/cons", OWRITE);
	/*
	 * init will reinitialize its namespace.
	 * #ec gets us plan9.ini settings (*var variables).
	 */
	bind("#ec", "/env", MREPL);
	bind("#e", "/env", MBEFORE|MCREATE);
	bind("#s", "/srv", MREPL|MCREATE);
#ifdef DEBUG
	print("argc=%d\n", argc);
	for(fd = 0; fd < argc; fd++)
		print("%#p %s ", argv[fd], argv[fd]);
	print("\n");
#endif DEBUG

	readfile("#e/cputype", cputype, sizeof(cputype));

	/*
	 *  set up usb keyboard, mouse and disk, if any.
	 */
	print("usbinit...");
	usbinit(Dontpost);

	/*
	 *  load keymap if it's there.
	 */
	print("kbmap...");
	kbmap();

	/*
	 * read disk partition tables here so that readnvram via factotum
	 * can see them.  ideally we would have this information in
	 * environment variables before attaching #S, which would then
	 * parse them and create partitions.
	 */
	rdparts = getenv("readparts");
	if(rdparts)
		readparts();
	free(rdparts);

	print("bind / /...");
	if(bind("/", "/", MREPL) < 0)
		fatal("bind /");
	setenv("rootdir", "/root");
	settime(1, -1, nil);
	swapproc();
	print("init...");
	init();

	print("init failed!!");
	sleep(10000);
	fatal("early init failed\n");
}

static void
swapproc(void)
{
	int fd;

	fd = open("#c/swap", OWRITE);
	if(fd < 0){
		warning("opening #c/swap");
		return;
	}
	if(write(fd, "start", 5) <= 0)
		warning("starting swap kproc");
	close(fd);
}

/*
static void
usbinit(void)
{
	static char usbd[] = "/boot/usbd";

	if(access("#u/usb/ctl", 0) >= 0 && bind("#u", "/dev", MAFTER) >= 0 &&
	    access(usbd, AEXIST) >= 0)
		run(usbd, nil);
}
*/

static void
kbmap(void)
{
	char *f;
	int n, in, out;
	char buf[1024];

	f = getenv("kbmap");
	if(f == nil)
		return;
	if(bind("#κ", "/dev", MAFTER) < 0){
		warning("can't bind #κ");
		return;
	}

	in = open(f, OREAD);
	if(in < 0){
		warning("can't open kbd map: %r");
		return;
	}
	out = open("/dev/kbmap", OWRITE);
	if(out < 0) {
		warning("can't open /dev/kbmap: %r");
		close(in);
		return;
	}
	while((n = read(in, buf, sizeof(buf))) > 0)
		if(write(out, buf, n) != n){
			warning("write to /dev/kbmap failed");
			break;
		}
	close(in);
	close(out);
}

/* End of boot.c code, beginning of init.c code */

void
init(void)
{
	int fd;
	char ctl[128];

	closefds();

	if(cpuflag)
		service = "cpu";
	else
		service = "terminal";
	manual = 0;

	snprint(ctl, sizeof(ctl), "#p/%d/ctl", getpid());
	fd = open(ctl, OWRITE);
	if(fd < 0)
		print("init: warning: can't open %s: %r\n", ctl);
	else
		if(write(fd, "pri 10", 6) != 6)
			print("init: warning: can't set priority: %r\n");
	close(fd);

	cpu = readenv("#e/cputype");
	if(access("#ec/service", 0) == 0){
		service = readenv("#ec/service");
	}
	if(access("#ec/manual", 0) == 0)
		checkmanual = readenv("#ec/manual");
	else
		checkmanual = strdup("noenv");
	if(access("#ec/cmd", 0) == 0)
		cmd = readenv("#ec/cmd");
	if(strcmp(checkmanual, "manual") == 0)
		manual = 1;
	settheenv("#e/objtype", cpu);
	settheenv("#e/service", service);
	print("service: %s...", service);

	print("bind # trees...");
	bind("#d", "/fd", MREPL);
	bind("#p", "/proc", MREPL);
	bind("#¤", "/dev", MAFTER);
	bind("#S", "/dev", MAFTER);
	bind("#l", "/net", MAFTER);
	bind("#I", "/net", MAFTER);
	bind("/boot", "/bin", MAFTER);

	if(manual == 0){
		print("plan9rc...\n");
		fexec(plan9start);
	}
	for(;;){
		print("\ninit: starting rc\n");
		fexec(rcexec);
		manual = 1;
		cmd = 0;
		sleep(1000);
	}
}

static int gotnote;

void
pinhead(void*, char *msg)
{
	gotnote = 1;
	fprint(2, "init got note '%s'\n", msg);
	noted(NCONT);
}

void
fexec(void (*execfn)(void))
{
	Waitmsg *w;
	int pid;

	switch(pid=fork()){
	case 0:
		rfork(RFNOTEG);
		(*execfn)();
		print("init: exec error: %r\n");
		exits("exec");
	case -1:
		print("init: fork error: %r\n");
		exits("fork");
	default:
	casedefault:
		notify(pinhead);
		gotnote = 0;
		w = wait();
		if(w == nil){
			if(gotnote)
				goto casedefault;
			print("init: wait error: %r\n");
			break;
		}
		if(w->pid != pid){
			free(w);
			goto casedefault;
		}
		if(strstr(w->msg, "exec error") != 0){
			print("init: exit string %s\n", w->msg);
			print("init: sleeping because exec failed\n");
			free(w);
			for(;;)
				sleep(1000);
		}
		if(w->msg[0])
			print("init: rc exit status: %s\n", w->msg);
		free(w);
		break;
	}
}

void
rcexec(void)
{
	if(cmd)
		execl("/boot/rc", "rc", "-c", cmd, nil);
	execl("/boot/rc", "rc", nil);
}

void
plan9start(void)
{
	if(rfork(RFPROC|RFNAMEG|RFFDG)==0)
		execl("/boot/rc", "rc", "-c", "/boot/plan9rc", nil);
	else
		wait();
		print("plan9rc completed, console is kernel only namespace\n");
		exits(nil);
}

char*
readenv(char *name)
{
	int f, len;
	Dir *d;
	char *val;

	f = open(name, OREAD);
	if(f < 0){
		print("init: can't open %s: %r\n", name);
		return "*unknown*";	
	}
	d = dirfstat(f);
	if(d == nil){
		print("init: can't stat %s: %r\n", name);
		return "*unknown*";
	}
	len = d->length;
	free(d);
	if(len == 0)	/* device files can be zero length but have contents */
		len = 64;
	val = malloc(len+1);
	if(val == nil){
		print("init: can't malloc %s: %r\n", name);
		return "*unknown*";
	}
	len = read(f, val, len);
	close(f);
	if(len < 0){
		print("init: can't read %s: %r\n", name);
		return "*unknown*";
	}else
		val[len] = '\0';
	return val;
}

void
settheenv(char *var, char *val)
{
	int fd;

	fd = create(var, OWRITE, 0644);
	if(fd < 0)
		print("init: can't open %s\n", var);
	else{
		fprint(fd, val);
		close(fd);
	}
}

void
cpenv(char *file, char *var)
{
	int i, fd;
	char buf[8192];

	fd = open(file, OREAD);
	if(fd < 0)
		print("init: can't open %s\n", file);
	else{
		i = read(fd, buf, sizeof(buf)-1);
		if(i <= 0)
			print("init: can't read %s: %r\n", file);
		else{
			close(fd);
			buf[i] = 0;
			settheenv(var, buf);
		}
	}
}

/*
 *  clean up after /boot
 */
void
closefds(void)
{
	int i;

	for(i = 3; i < 30; i++)
		close(i);
}

